src/Badges.sol

```diff
@@ -137,10 +153,31 @@ contract Badges is
     return balances[_owner];
   }
 
-  function ownerOf(uint256 _tokenId) external view virtual override returns (address) {
-    address owner_ = owners[_tokenId];
-    require(owner_ != address(0), "ownerOf: token doesn't exist");
-    return owner_;
+  function ownerOf(uint256 _tokenId) external view virtual override tokenExists(_tokenId) returns (address) {
+    return owners[_tokenId];
+  }
+
+  function revokeBadge(uint256 _raftTokenId, uint256 _badgeId)
+    external
+    tokenExists(_badgeId)
+    senderIsRaftOwner(_raftTokenId, "revokeBadge")
+  {
+    revokedBadgesHashes.set(_badgeId);
+    emit BadgeRevoked(_badgeId, msg.sender);
+  }
+
+  function reinstateBadge(uint256 _raftTokenId, uint256 _badgeId)
+    external
+    tokenExists(_badgeId)
+    senderIsRaftOwner(_raftTokenId, "reinstateBadge")
+  {
+    revokedBadgesHashes.unset(_badgeId);
+    emit BadgeReinstated(_badgeId, msg.sender);
+  }
+
+  function isBadgeValid(uint256 _badgeId) external view tokenExists(_badgeId) returns (bool) {
```

Comment:
and fuzz tests too..

Reply:
Not sure how fuzz tests would work here. For fuzzing you'll pass in a parameter (like a uint256) and it'll try a bunch of variations of that to try to break the function. What parameter would we use for fuzz tests?

Reply:
Regarding a test for isBadgeValid ... is a separate test needed? I wrote a test for revoking and reinstating which both use isBadgeValid ... wondering if we need it?

Reply:
Got your note about fuzz test - sounds good. 

for isBadgeValid, since it's an external function, let's have some test coverage there. In case we change the internal logic of revocation/expiration in the future, we need a test to get the confidence that it didn't impact the user